### PR TITLE
Granular code coverage with introspection

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add a new build option `SANITIZE_COVERAGE`. If it is enabled, the code is instrumented to track the coverage. The collected information is available inside ClickHouse with: (1) a new function `coverage` that returns an array of unique addresses in the code found after the previous coverage reset; (2) `SYSTEM RESET COVERAGE` query that resets the accumulated data. This allows us to compare the coverage of different tests, including differential code coverage. Continuation of #20539.

Example: https://pastila.nl/?00013e20/348ab745ec310eeb1994a3cca223db8e#2WCaoENmvHHK539A2WX6jQ==

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 70e3dd808cc3f087504892d18a9e61eb6f948151 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56102/70e3dd808cc3f087504892d18a9e61eb6f948151/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6680338399">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56102/70e3dd808cc3f087504892d18a9e61eb6f948151/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by alexey-milovidov** [comment]
I'm considering two implementation options:

1. Global coverage data. It will require running the tests sequentially, but it will give more complete information about what the server was doing during the test.
2. Per-thread coverage data, flushed into a system table. It will allow running the tests in parallel, but will only collect the data from threads attributed to queries (no data from background operations).

It looks like the first option is better for actual code coverage.

I'm also considering various details of the coverage data:
1. Unique addresses seen in the code - the most lightweight option.
2. Unique addresses seen in the code with the counter, how many times they were seen - for profiling.
3. Full trace, including every position in the code, possibly with timestamps. It is the heaviest option, but can be good for in-depth analysis of code behavior. But tracing every basic block is worse than tracing only large enough functions, see https://github.com/ClickHouse/ClickHouse/pull/34160

I'm also considering possible extensions of this instrumentation:
1. For example, we can allow dynamic fail-points for selected functions. But it will be better with https://github.com/ClickHouse/ClickHouse/pull/34160

**Comment by alexey-milovidov** [comment]
https://pastila.nl/?0046f9c8/c8bb84f34193de25c72efcc4a58ecc42#BN6lG7IT52u1TIKPPWZ1og==

**Comment by alexey-milovidov** [comment]
Another interesting example: https://pastila.nl/?00006624/d2e0f6a4abad5d74badcd274d84db8a6#5TT5XQ8sXXBV8KyzvVVyoA==

**Comment by alexey-milovidov** [comment]
I'm thinking about where to enable it - either it will be another build type (release+coverage) and another test run, or it will be included in the debug build.

**Comment by alexey-milovidov** [comment]
> It will require running the tests sequentially

We can run tests in parallel with coverage using multiple clickhouse-server processes.

**Comment by alexey-milovidov** [comment]
Example of unique lines covered by a test:
```
WITH '00647_histogram_negative.sql' AS name
SELECT addressToLine(arrayJoin(coverage) AS addr)
FROM system.coverage
WHERE (test_name = name) AND (addr NOT IN (
    SELECT arrayJoin(coverage) AS addr
    FROM system.coverage
    WHERE test_name != name
))

Query id: c3d198cd-4a93-4585-b746-2c5653f59bfe

┌─addressToLine(arrayJoin(coverage))──────────────────────────────────┐
│ ./build/./src/AggregateFunctions/AggregateFunctionHistogram.h:59    │
│ ./build/./contrib/llvm-project/libcxx/include/__algorithm/comp.h:73 │
│ ./build/./src/AggregateFunctions/AggregateFunctionHistogram.h:351   │
└─────────────────────────────────────────────────────────────────────┘

3 rows in set. Elapsed: 0.239 sec.
```

**Comment by alexey-milovidov** [comment]
Tests sorted by unique places in code, not covered by any other tests:
https://pastila.nl/?0002b5fa/a52ccac38845d436b6eb86baeb22e014#N51wuWHXYZ1kYNcBPth+EA==
(note: this is incomplete, the tests are still in progress)

**Comment by alexey-milovidov** [comment]
client, local, and other tools should dump the coverage into a file for subsequent insertion - we can control it with an environment variable.

**Comment by alexey-milovidov** [comment]
Functional tests can finish in several hours in sequential mode - this is passable.

**Comment by mkaynov** [comment]
> I'm thinking about where to enable it - either it will be another build type (release+coverage) and another test run, or it will be included in the debug build.

At this point, is the goal to have a real "test coverage map" in the CH database? For the sake of CI jobs simplicity I'd steer towards separate job (build/run). Maybe use it for master workflow only would be sufficient, but this of course depends on our needs.

File: tests/clickhouse-test [hunk 1]
```diff
@@ -1173,6 +1173,22 @@ class TestCase:
             description_full += result.reason.value
 
         description_full += result.description
+
+        if BuildFlags.SANITIZE_COVERAGE in args.build_flags:
```
Comment (Felixoid) [inline tests/clickhouse-test:1177]: Is `args` defined here?

*after 5 minutes of jumping around the code*

Oh well.. it's very obfuscated, where it's global, local, and where it has a completely different meaning than anywhere else in the file. But yes, it's definied globally in `if __name__ == "main"`

**Comment by alexey-milovidov** [comment]
Having a separate build job and check will be reasonable. It will be good if it runs for PRs because we can report if the coverage increased or decreased, the newly covered lines of code, etc. It will require functional integration and, possibly, unit tests. It could be nice to have it in Fuzzer and Stress tests, but for that purpose, we can reuse one of the existing jobs, e.g., replace release to release+coverage.

There are different goals for coverage:

1. Show a report with test coverage by files for a nice number (e.g. it was ~85% by lines by functional tests last time we collected a real report).

But it will require a few more steps:
- for attributing addresses to lines, we have to call `addressToLineWithInlines` for all addresses, and it could take +1 hour easily; `addressToLineWithInlines` because in the release build, the line in our code is often attributed to inlined code somewhere in, e.g. std::vector; we can also compile without inlining for coverage;
- the coverage can be calculated by "edges", "basic blocks", "functions", "lines", and "files"; currently, the code is instrumented on edges, which is more granular than basic blocks, and more granular than functions; but the usual report reports by lines; it is easy to visualize every edge or basic block by highlighting the line, where it starts, so it will be good for visualization, but for calculating the coverage by lines, we need to know how many lines every basic block / edge spans.

Note: if we calculate coverage by edges (there could be many unique edges on a single line due to template instantiations), it will be quite low, maybe 30%.

We can calculate the coverage by symbols (functions, taking different template instantiations differently).

2. Show some absolute or relative number to check if it grows or declines over time and in every pull request. Show the newly unique covered code and the new code that wasn't covered.

This will be easy to do.

3. Calculate statistical properties of tests and sort tests by their relevance to the modified lines of code. Enrich documentation by the information about tests - e.g. this function is covered by these tests in order of relevance (by the way, this is already possible by using the `used_functions` in the `query_log`).

This will be easy to do.

**Comment by alexey-milovidov** [comment]
Let's do this:

1. We can merge this PR that adds new build options and support for them in clickhouse-test.
2. I will do subsequent PRs for lightweight coverage in other tools (clickhouse-client, clickhouse-format, etc).
3. I will do subsequent PRs checking if we can enable coverage by default for some builds.

**Comment by alexey-milovidov** [comment]
> I will do subsequent PRs checking if we can enable coverage by default for some builds.

This: https://github.com/ClickHouse/ClickHouse/pull/58792

**Pull request merged**